decision_service/include/factory_resolver.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include "model_mgmt.h"
+#include "object_factory.h"
+
+namespace reinforcement_learning  {
+  using dtfactory = utility::object_factory<model_management::i_data_transport>;
+  using modelfactory = utility::object_factory<model_management::i_model>;
+
+  extern dtfactory& data_transport_factory;
```

Comment:
>extern dtfactory& data_transport_factory; [](start = 2, length = 41)

why extern?

Reply:
Every cpp module that needs to refer to data_transport_factory will include this file but it will only be instantiated in one cpp module.  

---
In reply to: [193405978](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193405978) [](ancestors = 193405978)

Reply:
so how does this work if we have multiple instances of the DS in the same process. Does the transport factory need to be the same? This looks like a big singleton too me.

---
In reply to: [193427006](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193427006) [](ancestors = 193427006,193405978)

Reply:
Yes that's right.  This is a c++ stdlib technique for singletons.    The intention is to allow users to register new transports, models etc before using client lib.

---
In reply to: [193484768](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193484768) [](ancestors = 193484768,193427006,193405978)

Reply:
I thought about what you said some more.  

I will add a constructor to rlclientlib that takes factories in addition to what's there.  When they are not provided, I will use the default factories.  This way the simple case is simple and the complex case is still possible.

---
In reply to: [193496818](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193496818) [](ancestors = 193496818,193484768,193427006,193405978)